### PR TITLE
[BEAM-8446] Retrying BQ query on timeouts

### PR DESCRIPTION
**Pull request by pabloem**
r: @udim 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python 3.7 PostCommit

**Comment by pabloem** [comment]
Run Python 3.7 PostCommit

**Comment by pabloem** [comment]
First instance of job running:
https://builds.apache.org/job/beam_PostCommit_Python37_PR/42/

**Comment by pabloem** [comment]
Run Python 3.7 PostCommit

**Comment by pabloem** [comment]
Run Python 3.7 PostCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
The build passed, but then timed out: https://builds.apache.org/job/beam_PostCommit_Python37_PR/44/console

**Comment by pabloem** [comment]
Run Python 3.7 PostCommit

**Comment by pabloem** [comment]
Again timeout without errors.

**Comment by pabloem** [comment]
Run Python 3.7 PostCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
Run Python 3.7 PostCommit

**Comment by pabloem** [comment]
Run Python 3.7 PostCommit

**Comment by pabloem** [comment]
Again timeout without failures: https://builds.apache.org/job/beam_PostCommit_Python37_PR/46/

File: sdks/python/apache_beam/io/gcp/tests/bigquery_matcher.py [hunk 1]
```diff
@@ -46,9 +47,11 @@
 MAX_RETRIES = 5
 
 
-def retry_on_http_and_value_error(exception):
+def retry_on_http_timeout_and_value_error(exception):
   """Filter allowing retries on Bigquery errors and value error."""
-  return isinstance(exception, (GoogleCloudError, ValueError))
+  return isinstance(exception, (GoogleCloudError,
+                                ValueError,
+                                concurrent.futures.TimeoutError))
```
Comment (udim) [inline sdks/python/apache_beam/io/gcp/tests/bigquery_matcher.py]: Where is this `TimeoutError` being raised? I didn't see it in the failed tests' logs .

**Comment by udim** [comment]
> Again timeout without failures: https://builds.apache.org/job/beam_PostCommit_Python37_PR/46/

I've tried looking at these with no luck. :(

**Comment by pabloem** [comment]
Hm I added it because the Pydoc specifies it is a possibility. I did not reproduce it because the job is flaky...
https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.QueryJob.html#google.cloud.bigquery.job.QueryJob.result

**Comment by pabloem** [comment]
@udim do you think it makes sense to add this?

**Comment by pabloem** [comment]
@udim

**Comment by pabloem** [comment]
@udim

**Comment by udim** [comment]
> Hm I added it because the Pydoc specifies it is a possibility. I did not reproduce it because the job is flaky...
> https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.QueryJob.html#google.cloud.bigquery.job.QueryJob.result

If it's listed then we should definitely handle it.

**Pull request merged**